### PR TITLE
handle postgres cancel requests

### LINKED ISSUE #9389: Support Cancelling requests in progress for PostgreSQL
### What
What would you like Teleport to do differently?
Add support for Cancelling requests in progress for PostgreSQL. https://goteleport.com/docs/database-access/faq/#which-postgresql-protocol-features-are-not-supported
### How
How would you implement this?

### Why
Why do you need this?
DBA's need to be able to cancel requests in progress in addition to other customer use cases. 
### Workaround
If a workaround exists, please include it.

### PR DESCRIPTION
**Pull request by GavinFrazar**
Closes #9389

This PR adds support for postgres cancel requests.

### CODE REVIEW

**Review by smallinsky — COMMENTED** [review]
First pass.

**Review by greedy52 — COMMENTED** [review]
First pass

FYI, seems RDS Proxy doesn't support cancel request XD so we are doing better
https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/rds-proxy.html

File: lib/srv/db/postgres/engine.go
Comment (GavinFrazar) [reply lib/srv/db/postgres/engine.go:118]: no audit log event. Since cancel requests establish a new connection, the only way to associate a cancel request with a session is by adding PID to the session as well, and there's no way to know if a cancel request actually cancelled anything either. And then every time the user hits `ctrl+c`, there will be audit log events for a new session start, cancel request sent, session end. I feel like there were other problems I'm forgetting that came up when discussing this with @smallinsky last week, but we opted to skip audit logging changes at least for now.

File: tool/tsh/db.go
Comment (GavinFrazar) [reply tool/tsh/db.go:653]: tunnel mode with auto reissue certs happens when you use `tsh proxy db --key-file=... --cert-file=...`. I'll add a comment

File: lib/utils/certs.go
Comment (GavinFrazar) [reply lib/utils/certs.go:295]: handshake never sets this field, so this is safe.

```
	// Leaf is the parsed form of the leaf certificate, which may be initialized
	// using x509.ParseCertificate to reduce per-handshake processing. If nil,
	// the leaf certificate will be parsed as needed.
```

File: lib/srv/alpnproxy/local_proxy.go
Comment (GavinFrazar) [reply lib/srv/alpnproxy/local_proxy.go:244]: prior to this change, the only thing that determined if certs were used to dial upstream (i.e. like `--tunnel`) was whether certs were set in the local proxy. Since we wanted to support non-tunnel postgres local proxy `tsh proxy db postgres`, I had to load certs from the profile in case tunnel was required for a cancel request. But then certs were loaded into local proxy but should only be used for cancel requests. So yes it's needed due to that requirement

File: lib/srv/db/postgres/test.go
Comment (GavinFrazar) [reply lib/srv/db/postgres/test.go:429]: no, I'll unexport

Comment (GavinFrazar) [reply lib/srv/db/postgres/test.go:429]: oh, actually yes we need to export `TestLongRunningQuery`. I'll just unexport `TestSecretKey`

File: lib/srv/db/postgres/proxy.go
Comment (GavinFrazar) [reply lib/srv/db/postgres/proxy.go:147]: I just didn't want to log the cancel request with `%#v` since that would log the secret key.

Comment (GavinFrazar) [reply lib/srv/db/postgres/proxy.go:147]: I'll add a patch that handles cancel request separately to remove the duplication.

File: lib/srv/db/postgres/engine.go
Comment (GavinFrazar) [reply lib/srv/db/postgres/engine.go:118]: sure

File: lib/srv/db/postgres/engine.go
Comment (GavinFrazar) [reply lib/srv/db/postgres/engine.go:528]: There is no need to `trace.Wrap` trace errors.

File: tool/tsh/db.go
Comment (GavinFrazar) [reply tool/tsh/db.go:1176]: note that this is in `withConnectRequirements`, as in `tsh db connect`. So this will only ensure we start  a local proxy with `tsh db connect`.

Comment (GavinFrazar) [reply tool/tsh/db.go:1176]: is that ok?

File: lib/srv/alpnproxy/local_proxy.go
Comment (GavinFrazar) [reply lib/srv/alpnproxy/local_proxy.go:428]: It's possible and I considered doing something like that, but it would require a lot more refactoring and complexity to do it. FWIW, all of the local proxy changes are due to the requirement that `tsh proxy db` (without `--tunnel` or `--cert-file/--key-file`) loads and uses certs for cancel requests, aka "tunnels on demand". If we instead just let cancel requests fail when client certs are not presented, then all of this extra complexity goes away. I consider it a bug that `psql` doesn't send cancel requests with SSL when I've configured `psql` to connect with client certs.

File: integration/appaccess/pack.go
Comment (GavinFrazar) [reply integration/appaccess/pack.go:388]: `CheckCertsNeeded` flag instead? I'm not sure what to name it. This comes from the requirement that `tsh proxy db` tunnels only as needed, whereas normally that command would not load and use client certs ever. If we instead allowed cancel requests to fail when not using `--tunnel`, then this complexity goes away. In my opinion, it is a client software error to send cancel requests without SSL when configured to use SSL (like `psql` does).

File: tool/tsh/db.go
Comment (smallinsky) [reply tool/tsh/db.go:1182]: > So this will only ensure we start a local proxy with tsh db connect.
Even if this is only `tsh db connect` the default behaviour was changed. If a customer was using Separate listener mode the `tsh db connect`  flow was not using local proxy. After that change if cluster was configured with separate_listener_mode  the `tsh db connect` will trigger TLS Routing flow. 
My main point is not to change this behaviour.

File: lib/srv/alpnproxy/local_proxy.go
Comment (greedy52) [reply lib/srv/alpnproxy/local_proxy.go:405]: Ya, refactoring can get complicated. The easiest i can think of is to let `middleware.OnNewConnection` return a per connection setting for overwriting downstreamConn and TLS certs. Still a lot of work to change existing middleware but the benefit would be to encapsulate cancel request stuff in a separate middleware that is only used when connecting to Postgres in non-tunnel mode. I'll leave it to you to decide if you are in the mood to refactor in some way or keep it as it is.

regardless, it would be great if we can move pgproto3 stuff to a separate util lib.

File: lib/srv/db/postgres/proxy.go
Comment (greedy52) [inline lib/srv/db/postgres/proxy.go:205]: good catch

File: lib/srv/db/postgres/test.go
Comment (greedy52) [inline lib/srv/db/postgres/test.go:390]: nit: should we delete even if cancel is nil?

**Review by greedy52 — COMMENTED** [review]
LTGM overall except the one Marek raised about `tsh db connect` usage. I agree that we shouldn't use local proxy if there is a separate port for Postgres.

Personally I would still prefer abstracting the postgres specific logic away from the local proxy, somehow, if possible. We added a few bool parameter a few places but that didn't eliminate checking against Postgres protocol in each level. Doesn't have to be now so maybe we can think about it a localproxy/middleware refactoring/redesign separately in the future to make this better

File: lib/srv/db/postgres/test.go
Comment (GavinFrazar) [reply lib/srv/db/postgres/test.go:390]: I think we might as well, although it doesn't really matter since PIDs are monotonically increasing to keep it simpler. I mainly wanted to prevent context leaks from failing to call cancel

File: tool/tsh/db.go
Comment (GavinFrazar) [reply tool/tsh/db.go:1182]: Is the concern here due to local proxy causing TLS-inside-TLS behavior? Because without this change, `tsh db connect` will not support cancel requests in separate listener mode.

Comment (smallinsky) [reply tool/tsh/db.go:1182]: > Is the concern here due to local proxy causing TLS-inside-TLS behavior? 
I don't that that TLS inside TLS is  our concern. We are using this behavior in other integration but we don't want to change/alter suddenly current behavior and make the `--tunnel` mode as default for all postgres connection. 
  
Why the proxy cannot run without --tunnel mode for normal connection and upgrade a new connection to TLS only if the cancel request was detected ?

Comment (smallinsky) [reply tool/tsh/db.go:1182]: @GavinFrazar 
Even if we upgrade the connection only to TLS without `--tunnel` option this behaviour can break the client integration. 

Let's imagine that the client is using the following sperate teleport proxy configuration: 
```
auth_service:
  proxy_listener_mode: "separate"
  
proxy_service:
  enabled: "true"
  public_addr:  ['alb.master.ice-berg.dev:3080']
  ssh_public_addr: ['master.ice-berg.dev:3023']
  tunnel_public_addr: ['master.ice-berg.dev:3024']
  postgres_public_addr: ['master.ice-berg.dev:5432']
  web_listen_addr: 0.0.0.0:3080
  postgres_listen_addr: 0.0.0.0:5432
```

where `alb.master.ice-berg.dev:`   is an L7 Load Balancer like AWS ALB that terminates TLS connection and dials the teleport proxy without forwarding TLS extensions. 

before this change the `tsh db connect postgres` command was using the `master.ice-berg.dev:3024` address after switching to a local proxy for postgres connection the `tsh db connect postgres` will try to connect to `alb.master.ice-berg.dev:3080` with TLS Routing extension. Due to L7 ALB the TLS exertion will be dropped and the client will not be able to connect to the right service. Recently this case was fixed by adding support for connection upgrade but I can still imagine that the v13 client can be used to connect to the old teleport proxy that doesn't support [Connection Upgrade](https://github.com/gravitational/teleport/issues/14369) feature.

Comment (greedy52) [reply tool/tsh/db.go:1182]: Hey guys. I know we have code to track what mode is required, when, which protocol, etc.. but I've decided to create a spreadsheet to track it as it's slightly more visual. Feel free to update it or fix it if something is not accurate:
https://docs.google.com/spreadsheets/d/1m0swp5F4PZ9-C6pAUR45FbIB1o2IeJ3mjmJz1VrRG3g/edit?usp=sharing

**Comment by GavinFrazar** [comment]
@smallinsky @greedy52 updated to move the postgres local proxy requirement to only when TLS routing is enabled per marek's suggestion, please take another look

**Comment by GavinFrazar** [comment]
This local proxy requirement for postgres with TLS routing mode is wrong I think.

Local proxy postgres with TLS routing should only require a local proxy when the proxy is behind an L7 load balancer.

**Comment by GavinFrazar** [comment]
@smallinsky `tsh db connect` will always use a local proxy when TLS routing is enabled. That's here:
```
func withConnectRequirements(ctx context.Context, tc *client.TeleportClient, route *tlsca.RouteToDatabase) requireOpt {
	return func(r *dbLocalProxyRequirement) {
		if !r.localProxy && tc.TLSRoutingEnabled {
			r.addLocalProxy(formatTLSRoutingReason(tc.SiteName))
		}
```

What I'm saying is that adding a requirement for all the other commands in `getDBLocalProxyRequirement` is wrong. As you mention, users can connect directly without local proxy, so `tsh db config` and `tsh db env` should still for postgres with TLS routing enabled.

So I think I should revert the change in `getDBLocalProxyRequirement` here:
https://github.com/gravitational/teleport/pull/22173/commits/dd2f5f5ad64370cd6cd62dd98bd7ada80b89f739#diff-dae87d4d3799335d85a5fa6ff17cfcbfb30705ae145c9e67ef70286423676a48R1153-R1154

**Comment by GavinFrazar** [comment]
The broader issues with L7 load balancer and other db protocols need to be addressed in another PR. We should probably make use of `IsALPNConnUpgradeRequired` (and maybe save that info in the user's .tsh profile) for that purpose.

I think this PR is otherwise ready to merge, if you approve of removing local proxy requirement for postgres with TLS routing in `getDBLocalProxyRequirement` then we can merge this.
```diff
	case defaults.ProtocolMySQL,
-		defaults.ProtocolPostgres:
		if tc.TLSRoutingEnabled {
			out.addLocalProxy(fmt.Sprintf("%v and %v",
				formatDBProtocolReason(route.Protocol),
				formatTLSRoutingReason(tc.SiteName)))
		}
```

**Comment by greedy52** [comment]
> The broader issues with L7 load balancer and other db protocols need to be addressed in another PR. We should probably make use of IsALPNConnUpgradeRequired (and maybe save that info in the user's .tsh profile) for that purpose.

I will work on this next week to address it separately. And agreed we shouldn't require local proxy for Postgres for `tsh db env/config` when TLS routing is enabled.

**Review by greedy52 — APPROVED** [review]

**Review by smallinsky — APPROVED** [review]

**Pull request merged**
